include/pci.h

```diff
@@ -29,36 +31,74 @@ extern const char *pci_class_code[];
 #define PCI_BAR_IO_MASK 3
 #define PCI_BAR_MEMORY_MASK 15
 
+#define PCIR_DEVICEID 0x00
+#define PCIR_VENDORID 0x02
+#define PCIR_STATUS 0x04
+#define PCIR_COMMAND 0x06
+#define PCIR_CLASSCODE 0x08
+#define PCIR_IRQPIN 0x3e
+#define PCIR_IRQLINE 0x3f
+#define PCIR_BAR(i) (0x10 + (i)*4)
+
 typedef struct pci_device pci_device_t;
+typedef TAILQ_HEAD(, pci_device) pci_device_list_t;
+
+typedef struct pci_addr {
+  uint8_t bus;
+  uint8_t device;
+  uint8_t function;
+} pci_addr_t;
+
+typedef uint32_t (*pci_read_config_t)(pci_device_t *device, unsigned reg,
+                                      unsigned size);
+typedef void (*pci_write_config_t)(pci_device_t *device, unsigned reg,
+                                   unsigned size, uint32_t value);
+
+typedef struct pci_bus {
+  pci_read_config_t read_config;
+  pci_write_config_t write_config;
+} pci_bus_t;
 
-typedef struct {
-  pm_addr_t addr;
-  size_t size;
-  pci_device_t *dev; /* Reference to the parent device. */
-  int i;             /* This BAR no in parent device. */
-} pci_bar_t;
-
-typedef struct pci_device {
-  struct {
-    uint8_t bus;
-    uint8_t device;
-    uint8_t function;
-  } addr;
+#define PCI_BUS_DECLARE(name) extern pci_bus_t name[1]
+
+struct pci_device {
+  TAILQ_ENTRY(pci_device) link;
+
+  pci_bus_t *bus;
+  pci_addr_t addr;
 
   uint16_t device_id;
   uint16_t vendor_id;
   uint8_t class_code;
   uint8_t pin, irq;
 
   unsigned nbars;
-  pci_bar_t bar[6];
-} pci_device_t;
-
-typedef struct {
-  unsigned ndevs;
-  pci_device_t *dev;
-} pci_bus_t;
+  resource_t bar[6];
+};
+
+typedef struct pci_bus_device {
```

Comment:
AFAIU eventually `pci_bus_device` will become particular instance of `device` type and `pci_bus` will become special case of `bus` type. To put it differently `pci_bus_device` holds the state of device that happens to be a PCI bus, while `pci_bus` delivers common set of functions to operate on the bus.